File: batou-rules/testdata/fixtures/java/vulnerable/LogInjection.java
package com.example.vulnerable;

import java.io.IOException;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class LogInjection extends HttpServlet {

    private static final Logger logger = LoggerFactory.getLogger(LogInjection.class);

    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String username = request.getParameter("username");
        String action = request.getParameter("action");

        logger.info("User login attempt: " + request.getParameter("username") + " action=" + action);

        if (authenticate(username, request.getParameter("password"))) {
            logger.info("Authentication successful for user: " + username);
            response.sendRedirect("/dashboard");
        } else {
            logger.warn("Failed login for user: " + username + " from IP: " + request.getRemoteAddr());
            response.sendError(401, "Invalid credentials");
        }
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String searchTerm = request.getParameter("q");
        logger.debug("Search query from user: " + searchTerm);
        System.out.println("Search performed: " + request.getParameter("q"));
    }

    private boolean authenticate(String username, String pass) {
        return false;
    }
}

File: batou-rules/testdata/fixtures/kotlin/vulnerable/LogInjection.kt
package com.example.vulnerable

import javax.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import timber.log.Timber

private val logger = LoggerFactory.getLogger("auth")

class LogInjectionController {

    // Vulnerable: SLF4J Logger.warn with user-controlled concatenation (CWE-117)
    fun onLogin(req: HttpServletRequest) {
        val user = req.getParameter("user")
        logger.warn("login attempt: " + user)
    }

    // Vulnerable: Logger.debug with user input
    fun onDebug(req: HttpServletRequest) {
        val user = req.getParameter("user")
        logger.debug("processing: " + user)
    }

    // Vulnerable: Logger.trace with user input
    fun onTrace(req: HttpServletRequest) {
        val user = req.getParameter("user")
        logger.trace("trace: " + user)
    }

    // Vulnerable: System.out.println with tainted data
    fun onStdout(req: HttpServletRequest) {
        val user = req.getParameter("user")
        System.out.println("user: " + user)
    }

    // Vulnerable: System.err.println with tainted data
    fun onStderr(req: HttpServletRequest) {
        val user = req.getParameter("user")
        System.err.println("error for: " + user)
    }

    // Vulnerable: Jake Wharton's Timber with tainted data
    fun onTimber(req: HttpServletRequest) {
        val user = req.getParameter("user")
        Timber.d("debug user: " + user)
        Timber.e("error user: " + user)
        Timber.wtf("fatal: " + user)
    }

    // Vulnerable: String.format with user-controlled format string (CWE-134)
    fun onFormat(req: HttpServletRequest) {
        val fmt = req.getParameter("fmt")
        val msg = String.format(fmt)
        logger.info(msg)
    }
}
